lib/cdo/firehose.rb

```diff
@@ -19,50 +21,73 @@
+  BYTES_PER_RECORD = 1024 * 1000
+
+  # 'The PutRecordBatch operation can take up to 500 records per call or 4 MiB per call, whichever is smaller.'
+  # Ref: https://docs.aws.amazon.com/firehose/latest/dev/limits.html
+  BYTES_PER_REQUEST = 1024 * 1024 * 4
 
   # Initializes the @firehose to an AWS Firehose client.
-  def initialize
-    if [:development, :test].include? rack_env
-      return
+  def initialize(log: CDO.log)
+    super(
+      batch_count: ITEMS_PER_REQUEST,
+      batch_size: BYTES_PER_REQUEST,
+      max_interval: 10.0,
+      min_interval: 0.1,
+      log: log
+    )
+    unless [:development, :test].include? rack_env
+      self.client = Aws::Firehose::Client.new
     end
-    @firehose = Aws::Firehose::Client.new(region: REGION)
   end
 
   # Posts a record to the analytics stream.
   # @param data [hash] The data to insert into the stream.
   def put_record(data)
     return unless Gatekeeper.allows('firehose', default: true)
-
-    data_with_common_values = add_common_values(data)
-
-    if [:development, :test].include? rack_env
-      CDO.log.info "Skipped sending record to #{STREAM_NAME}: "
-      CDO.log.info data
-      return
+    record = add_common_values(data).to_json
+    if (size = record.bytesize) > BYTES_PER_RECORD
+      raise ArgumentError.new("Record too large (#{size} bytes)")
     end
+    buffer(record)
+  end
 
-    # TODO(asher): Determine whether these should be cached and batched via
-    # put_record_batch. See
-    #   http://docs.aws.amazon.com/sdkforruby/api/Aws/Firehose/Client.html#put_record_batch-instance_method
-    # for documentation.
-    @firehose.put_record(
-      {
+  def flush(records)
+    return unless Gatekeeper.allows('firehose', default: true)
+    if client
+      client.put_record_batch(
         delivery_stream_name: STREAM_NAME,
-        record: {data: data_with_common_values.to_json}
-      }
-    )
-  # Swallow and log all errors because an issue sending analytics should not prevent the caller from continuing.
-  rescue StandardError => error
-    # TODO(suresh): if the exception is Firehose ServiceUnavailableException, we should consider
-    # backing off and retrying.
-    # See http://docs.aws.amazon.com/sdkforruby/api/Aws/Firehose/Client.html#put_record-instance_method.
-    Honeybadger.notify(error)
+        records: records.map {|r| {data: r}}
+      )
+    else
+      log.info "Skipped sending records to #{STREAM_NAME}:\n#{records}"
+    end
+  end
+
+  REQUEST_OVERHEAD = {DeliveryStreamName: STREAM_NAME, Records: []}.to_json.size
+  RECORD_OVERHEAD = {Data: ''}.to_json.size
+
+  # Calculate the exact request size of a PutRecordBatch call given the provided records.
+  def size(records)
+    REQUEST_OVERHEAD + records.sum do |r|
+      RECORD_OVERHEAD +
+        # Base64-converted data is 4/3 the size of the original content.
+        4 * (r.bytesize / 3.to_f).ceil +
+        # Comma separating Records in the array.
+        1
+    end - 1
```

Comment:
Maybe leave a comment about why there is a `-1`. I'm guessing its for the fencepost issue from the `+1` of the commas.